### PR TITLE
Phase_ramp + line_filter

### PR DESCRIPTION
**Pull request by FENGSHAN95**
Dear Alex,

I created two functions: phase_ramp and line_filter_function. No test files so far.

I didn't finish line_filter_function. I got stuck with the rotated rectangle. Don't know how to assign the chosen area to the image. Maybe there should be some smart way. You may have some good ideas.

## Contributor Checklist:

* [ ] I have added or updated my entry in [the creators.json file](https://github.com/LiberTEM/LiberTEM/blob/master/packaging/creators.json)
* [ ] I have added [a changelog entry](https://github.com/LiberTEM/LiberTEM/tree/master/docs/source/changelog) for my contribution
* [ ] I have added/updated documentation for all user-facing changes
* [ ] I have added/updated test cases
* [ ] I have included the [rebuilt production build of the client](https://libertem.github.io/LiberTEM/contributing.html?#building-the-client) (only if
changes were made to the GUI)

## Please remove this section

Starting by submitting an incomplete pull request (PR) or draft PR is OK. You
can work on the checklist step by step by pushing additional commits into the
PR. Please indicate if you think some items may not be applicable.

You can have a look at [our contributing
docs](https://libertem.github.io/LiberTEM/contributing.html) for more
information on contributing to LiberTEM. Please feel free to ask for
clarification and help, for example in your PR description, with comments or in
our [Gitter channel](https://gitter.im/LiberTEM/Lobby).

Thank you for your contribution!

### CODE REVIEW

**Comment by sk1p** [comment]
Hi, and thanks for the contribution!

> I created two functions: phase_ramp and line_filter_function. No test files so far.

I think we can work on the tests when the functionality is there. Then, as a first step, the [test failures](https://travis-ci.com/github/LiberTEM/LiberTEM/builds/185666620) need to be fixed - this should be possible without adding a new test. All test environments, except for `mypy`, should have a green check-mark.

> I didn't finish line_filter_function. I got stuck with the rotated rectangle. Don't know how to assign the chosen area to the image. Maybe there should be some smart way. You may have some good ideas.

Without reinventing the wheel, maybe you can use [`skimage.draw.polygon`](https://scikit-image.org/docs/dev/api/skimage.draw.html#skimage.draw.polygon), and rotate by transforming the vertices?

As a general idea, maybe finding and correcting the phase ramp could be decoupled into two separate functions - this would make it possible to perform aggressive filtering (for example, low-pass) before looking at the minimum/maximum, and then remove the ramp from the unfiltered hologram. What do you think?

File: src/libertem/udf/holography.py [hunk 1]
```diff
@@ -69,6 +73,75 @@ def aperture_function(r, apradius, rsmooth):
...
+        ramp_x, ramp_y = ramp
+        mid_pos = img_size / 2 
+
+    if order==1:
+        img_new = np.zeros_like(img)
+        for i in np.arange(0, img_size[0], 1):
+            for j in np.arange(0, img_size[1], 1):
+                img_new[i, j] = img[i, j] + ramp_x * (mid_pos[0] / 2 - i) + ramp_y * (mid_pos[1] / 2 - j)
+    else:
+        # To be expanded.
```
Comment (sk1p) [inline src/libertem/udf/holography.py:140]: If you add a `pass`, or if you raise an `Exception` here, the syntax should be correct and the existing tests should run. Let me know if you need help with running the tests on your machine.

File: src/libertem/udf/holography.py [hunk 2]
```diff
@@ -69,6 +73,75 @@ def aperture_function(r, apradius, rsmooth):
...
+        ramp_y = (max_phase - min_phase) / (max_pos[1] - min_pos[1])
+        mid_pos = min_pos + max_pos
+    else:
+        ramp_x, ramp_y = ramp
+        mid_pos = img_size / 2 
+
+    if order==1:
+        img_new = np.zeros_like(img)
+        for i in np.arange(0, img_size[0], 1):
+            for j in np.arange(0, img_size[1], 1):
```
Comment (sk1p) [inline src/libertem/udf/holography.py:137]: As discussed, it should be possible to speed up these loops using numba. Did you benchmark this already?

Comment (FENGSHAN95) [reply src/libertem/udf/holography.py:142]: Yes. I update this. Have a check.

File: src/libertem/udf/holography.py (hunk 1 above)
Comment (FENGSHAN95) [reply src/libertem/udf/holography.py:145]: Added.

**Comment by FENGSHAN95** [comment]
> Hi, and thanks for the contribution!
> 
> > I created two functions: phase_ramp and line_filter_function. No test files so far.
> 
> I think we can work on the tests when the functionality is there. Then, as a first step, the [test failures](https://travis-ci.com/github/LiberTEM/LiberTEM/builds/185666620) need to be fixed - this should be possible without adding a new test. All test environments, except for `mypy`, should have a green check-mark.
> 
> > I didn't finish line_filter_function. I got stuck with the rotated rectangle. Don't know how to assign the chosen area to the image. Maybe there should be some smart way. You may have some good ideas.
> 
> Without reinventing the wheel, maybe you can use [`skimage.draw.polygon`](https://scikit-image.org/docs/dev/api/skimage.draw.html#skimage.draw.polygon), and rotate by transforming the vertices?
> 
> As a general idea, maybe finding and correcting the phase ramp could be decoupled into two separate functions - this would make it possible to perform aggressive filtering (for example, low-pass) before looking at the minimum/maximum, and then remove the ramp from the unfiltered hologram. What do you think?

**Comment by FENGSHAN95** [comment]
> Hi, and thanks for the contribution!
> 
> > I created two functions: phase_ramp and line_filter_function. No test files so far.
> 
> I think we can work on the tests when the functionality is there. Then, as a first step, the [test failures](https://travis-ci.com/github/LiberTEM/LiberTEM/builds/185666620) need to be fixed - this should be possible without adding a new test. All test environments, except for `mypy`, should have a green check-mark.
> 
> > I didn't finish line_filter_function. I got stuck with the rotated rectangle. Don't know how to assign the chosen area to the image. Maybe there should be some smart way. You may have some good ideas.
> 
> Without reinventing the wheel, maybe you can use [`skimage.draw.polygon`](https://scikit-image.org/docs/dev/api/skimage.draw.html#skimage.draw.polygon), and rotate by transforming the vertices?
> 
> As a general idea, maybe finding and correcting the phase ramp could be decoupled into two separate functions - this would make it possible to perform aggressive filtering (for example, low-pass) before looking at the minimum/maximum, and then remove the ramp from the unfiltered hologram. What do you think?

I update line filter following your suggestion. Please have a check.
I think it is a good idea to separate finding and correcting phase_ramp. Now this function is very basic. Still a lot needs to improve.

**Comment by uellue** [comment]
@FENGSHAN95 if you run the tests locally before pushing, you can get a faster turn-around. See https://libertem.github.io/LiberTEM/contributing.html#running-the-tests for instructions!

**Comment by FENGSHAN95** [comment]
> @FENGSHAN95 if you run the tests locally before pushing, you can get a faster turn-around. See https://libertem.github.io/LiberTEM/contributing.html#running-the-tests for instructions!

I will have a look. Sorry that I am not that familiar with coding.

**Comment by uellue** [comment]
@FENGSHAN95 do you have a local example to test your changes? Some test data and a test notebook?

**Comment by FENGSHAN95** [comment]
> @FENGSHAN95 do you have a local example to test your changes? Some test data and a test notebook?

Yes. I have.

**Comment by uellue** [comment]
@FENGSHAN95 Your change causes a syntax error in CI: https://travis-ci.com/github/LiberTEM/LiberTEM/jobs/389291762#L279. Why doesn't it happen if you run the same code locally using your test data and a test notebook? Can you make sure that your changed code is actually executed by your test notebook, for example with a `print()` statement?

**Comment by FENGSHAN95** [comment]
> @FENGSHAN95 Your change causes a syntax error in CI: https://travis-ci.com/github/LiberTEM/LiberTEM/jobs/389291762#L279. Why doesn't it happen if you run the same code locally using your test data and a test notebook? Can you make sure that your changed code is actually executed by your test notebook, for example with a `print()` statement?

Thank Alex. Now most of the errors should be fixed.

**Comment by sk1p** [comment]
I'm going to close this as it was moved to https://github.com/LiberTEM/LiberTEM-holo

**Pull request closed without merging**